File: bench/python/ops/bench_gaussian.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA Gaussian operator benchmark - Python equivalent of BenchGaussian.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_dtype_size,
    get_num_channels,
    get_format_from_dtype,
    get_border_type,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def gaussian(state):
    """Gaussian operator benchmark matching C++ BenchGaussian.cpp"""

    N, H, W = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    sigma = state.get_float64("sigma")
    border = get_border_type(state.get_string("border"))
    input_kind = get_input_kind(state.get_string("inputKind"))
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"
    device_id = state.get_device()
    nc = get_num_channels(dtype_str)

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip("Gaussian benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts")
        return None

    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind == "VarShape":
        state.skip("Fake-planar (NCHW_FAKE) Gaussian benchmark is tensor-only")
        return None

    mult = 3 if get_dtype(dtype_str) == cvcuda.Type.U8 else 4
    kernel_size = round(sigma * mult * 2 + 1)
    if kernel_size % 2 == 0:
        kernel_size += 1

    dtype_size = get_dtype_size(dtype_str)
    bytes_ = N * H * W * dtype_size
    if is_fake_planar:
        state.add_global_memory_reads(3 * bytes_)
        state.add_global_memory_writes(3 * bytes_)
    else:
        state.add_global_memory_reads(bytes_)
        state.add_global_memory_writes(bytes_)

    get_stream = create_stream_cache()

    if is_fake_planar:
        src = create_tensor(
            (N, nc, H, W), dtype_str, device_id, layout="NCHW", fill_mode="checkerboard"
        )
        inter_src = create_tensor(
            (N, H, W, nc), dtype_str, device_id, layout="NHWC", fill_mode=0
        )
        inter_dst = create_tensor(
            (N, H, W, nc), dtype_str, device_id, layout="NHWC", fill_mode=0
        )
        dst = create_tensor(
            (N, nc, H, W), dtype_str, device_id, layout="NCHW", fill_mode=0
        )

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            cvcuda.gaussian_into(
                inter_dst,
                inter_src,
                kernel_size=(kernel_size, kernel_size),
                sigma=(sigma, sigma),
                border=border,
                stream=stream,
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    if input_kind == "Tensor":  # Tensor mode
        tensor_shape = (N, nc, H, W) if is_planar else (N, H, W, nc)
        src = create_tensor(
            tensor_shape, dtype_str, device_id, layout=layout, fill_mode="checkerboard"
        )
        dst = create_tensor(
            tensor_shape, dtype_str, device_id, layout=layout, fill_mode=0
        )

        def run(launch):
            cvcuda.gaussian_into(
                dst,
                src,
                kernel_size=(kernel_size, kernel_size),
                sigma=(sigma, sigma),
                border=border,
                stream=get_stream(launch),
            )

        return run
    else:  # ImageBatchVarShape mode
        if is_planar and get_dtype(dtype_str) == cvcuda.Type.U8 and nc == 4:
            state.skip(
                "Gaussian RGBA8p planar var-shape is unsupported by the Python image API"
            )
            return None

        img_format = get_format_from_dtype(dtype_str, nc, planar=is_planar)

        src = create_image_batch_varshape(
            (N, H, W, nc),
            0,
            img_format,
            dtype_str,
            device=device_id,
            fill_mode="checkerboard",
        )
        dst = create_image_batch_varshape(
            (N, H, W, nc), 0, img_format, dtype_str, device=device_id, fill_mode=0
        )
        kernel_size_tensor = create_tensor(
            (N, 2),
            cvcuda.Type.S32,
            device_id,
            layout="NC",
            fill_mode=(kernel_size, kernel_size),
        )
        sigma_tensor = create_tensor(
            (N, 2), cvcuda.Type.F64, device_id, layout="NC", fill_mode=(sigma, sigma)
        )

        def run(launch):
            cvcuda.gaussian_into(
                src=src,
                dst=dst,
                max_kernel_size=(kernel_size, kernel_size),
                kernel_size=kernel_size_tensor,
                sigma=sigma_tensor,
                border=border,
                stream=get_stream(launch),
            )

        return run


if __name__ == "__main__":
    run_benchmark("gaussian", gaussian)

File: bench/cpp/ops/BenchGaussian.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchGaussianConfig.hpp"

#include <cvcuda/OpGaussian.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

#include <string_view>

namespace {

enum class GaussianLayout
{
    NHWC,
    NCHW,
    NCHWFake
};

GaussianLayout GetGaussianLayout(std::string_view layout, benchutils::InputKind inputKind)
{
    if (layout == "NHWC")
    {
        return GaussianLayout::NHWC;
    }
    if (layout == "NCHW")
    {
        return GaussianLayout::NCHW;
    }
    if (layout == "NCHW_FAKE")
    {
        if (inputKind == benchutils::InputKind::VarShape)
        {
            throw std::invalid_argument("Fake-planar (NCHW_FAKE) Gaussian benchmark is tensor-only");
        }
        return GaussianLayout::NCHWFake;
    }

    throw std::invalid_argument("Gaussian benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
}

bool IsPlanar(GaussianLayout layout)
{
    return layout == GaussianLayout::NCHW;
}

bool IsFakePlanar(GaussianLayout layout)
{
    return layout == GaussianLayout::NCHWFake;
}

void AddGaussianMemoryTraffic(nvbench::state &state, long bytes, GaussianLayout layout)
{
    if (IsFakePlanar(layout))
    {
        state.add_global_memory_reads(3 * bytes);
        state.add_global_memory_writes(3 * bytes);
    }
    else
    {
        state.add_global_memory_reads(bytes);
        state.add_global_memory_writes(bytes);
    }
}

template<typename T>
void ValidatePlanarChannels(GaussianLayout layout)
{
    if (!IsPlanar(layout))
    {
        return;
    }

    constexpr int ch = nvcv::cuda::NumElements<T>;
    if constexpr (ch == 2)
    {
        throw std::invalid_argument("Gaussian planar benchmark does not support 2-channel formats");
    }
    if constexpr (ch == 1)
    {
        throw std::invalid_argument("Gaussian planar benchmark requires multi-channel planar formats");
    }
}

template<typename T>
nvcv::Tensor MakeGaussianTensor(int3 shape, GaussianLayout layout)
{
    using BT         = typename nvcv::cuda::BaseType<T>;
    constexpr int ch = nvcv::cuda::NumElements<T>;

    if (IsPlanar(layout))
    {
        return nvcv::Tensor(
            nvcv::TensorShape{
                {shape.x, ch, shape.y, shape.z},
                "NCHW"
        },
            benchutils::GetDataType<BT>());
    }

    return nvcv::Tensor(
        nvcv::TensorShape{
            {shape.x, shape.y, shape.z, ch},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
}

template<typename T>
void FillGaussianImageBatches(nvcv::ImageBatchVarShape &src, nvcv::ImageBatchVarShape &dst, int3 shape,
                              GaussianLayout layout)
{
    if (IsPlanar(layout))
    {
        benchutils::FillPlanarImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0});
        benchutils::FillPlanarImageBatchLike<T>(dst, src);
    }
    else
    {
        benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0}, benchutils::CheckerboardValues<T>());
        benchutils::FillImageBatchLike<T>(dst, src, [](const long4_16a &) { return T{0}; });
    }
}

} // namespace

template<typename T>
inline void gaussian(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape     = benchutils::GetShape<3, int3>(state.get_string("shape"));
    auto                        layoutStr = benchutils::GetStringParam(state, "layout", "NHWC");
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));
    double                      sigma     = state.get_float64("sigma");

    NVCVBorderType borderType = benchutils::GetBorderType(state.get_string("border"));

    GaussianLayout layout = GetGaussianLayout(layoutStr, inputKind);

    int  kernelSize = (int)std::round(sigma * (std::is_same_v<nvcv::cuda::BaseType<T>, uint8_t> ? 3 : 4) * 2 + 1) | 1;
    int2 ksize2{kernelSize, kernelSize};

    nvcv::Size2D kernelSize2{kernelSize, kernelSize};
    double2      sigma2{sigma, sigma};

    const long bytes = static_cast<long>(shape.x) * shape.y * shape.z * sizeof(T);
    AddGaussianMemoryTraffic(state, bytes, layout);

    cvcuda::Gaussian op(kernelSize2, shape.x);

    // clang-format off

    if (IsFakePlanar(layout)) // tensor-only: planar->interleaved->Gaussian->interleaved->planar
    {
        using BT = typename nvcv::cuda::BaseType<T>;
        int  ch  = nvcv::cuda::NumElements<T>;

        nvcv::Tensor src     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interSrc({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interDst({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        cvcuda::Reformat reformatOp;

        benchutils::warmup_and_exec(state, BENCH_GAUSSIAN_WARMUP_ITERATIONS,
            [&op, &reformatOp, &src, &interSrc, &interDst, &dst, &kernelSize2, &sigma2, &borderType](cudaStream_t s) {
                reformatOp(s, src, interSrc);                                      // NCHW -> NHWC
                op(s, interSrc, interDst, kernelSize2, sigma2, borderType);         // interleaved Gaussian
                reformatOp(s, interDst, dst);                                      // NHWC -> NCHW
            });
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        using BT = typename nvcv::cuda::BaseType<T>;

        ValidatePlanarChannels<T>(layout);
        nvcv::Tensor src = MakeGaussianTensor<T>(shape, layout);
        nvcv::Tensor dst = MakeGaussianTensor<T>(shape, layout);

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        benchutils::warmup_and_exec(state, BENCH_GAUSSIAN_WARMUP_ITERATIONS,
            [&op, &src, &dst, &kernelSize2, &sigma2, &borderType](cudaStream_t s) {
                op(s, src, dst, kernelSize2, sigma2, borderType);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        ValidatePlanarChannels<T>(layout);

        nvcv::ImageBatchVarShape src(shape.x);
        nvcv::ImageBatchVarShape dst(shape.x);
        FillGaussianImageBatches<T>(src, dst, shape, layout);

        nvcv::Tensor kernelSizeTensor({{shape.x}, "N"}, nvcv::TYPE_2S32);
        nvcv::Tensor sigmaTensor({{shape.x}, "N"}, nvcv::TYPE_2F64);

        benchutils::FillTensor<int2>(kernelSizeTensor, [&ksize2](const long4_16a &){ return ksize2; });
        benchutils::FillTensor<double2>(sigmaTensor, [&sigma2](const long4_16a &){ return sigma2; });

        benchutils::warmup_and_exec(state, BENCH_GAUSSIAN_WARMUP_ITERATIONS,
            [&op, &src, &dst, &kernelSizeTensor, &sigmaTensor, &borderType](cudaStream_t s) {
                op(s, src, dst, kernelSizeTensor, sigmaTensor, borderType);
            });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(gaussian, NVBENCH_TYPE_AXES(BENCH_GAUSSIAN_TYPES))
BENCH_GAUSSIAN_AXES;
